File: packages/compiler/src/gen_circom.ts
type Graph = {
    type: string;
    edges: Record<string, number>;
}[];

function genCircomAllstr(
  graph_json: Graph,
  template_name: string,
  regex_str = ""
): string {
  /**
   * This function generates a Circom circuit from a given graph_json, template_name, and regex_str.
   * @param {Object} graph_json - The graph in JSON format.
   * @param {string} template_name - The name to be used for the Circom template.
   * @param {string} regex_str - The regular expression string, used only to print in a comment at the top.
   */

  const N = graph_json.length;
  // console.log(JSON.stringify(graph_json, null, 2));
  // const graph = Array(N).fill({});
  const rev_graph: Record<number, Record<number, number[]>> = {};
  const to_init_graph: number[][] = [];
  let init_going_state: number | null = null;

  for (let i = 0; i < N; i++) {
    rev_graph[i] = {};
    to_init_graph.push([]);
  }

  const accept_nodes: Set<number> = new Set();
  for (let i = 0; i < N; i++) {
    const node = graph_json[i];
    for (let k in node.edges) {
      const v: number = node.edges[k];
      rev_graph[v][i] = Array.from(JSON.parse(k)).map((c) =>
        (c as string).charCodeAt(0)
      );
      if (i === 0) {
        const index = rev_graph[v][i].indexOf(94);
        if (index !== -1) {
          init_going_state = v;
          rev_graph[v][i][index] = 255;
        }
        for (let j = 0; j < rev_graph[v][i].length; j++) {
          if (rev_graph[v][i][j] == 255) {
            continue;
          }
          to_init_graph[v].push(rev_graph[v][i][j]);
        }
      }
    }
    if (node.type == "accept") {
      accept_nodes.add(i);
    }
  }

  if (init_going_state !== null) {
    for (const [going_state, chars] of Object.entries(to_init_graph)) {
      const going_state_num = Number(going_state);
      if (chars.length === 0) {
        continue;
      }
      if (rev_graph[going_state_num][init_going_state] == null) {
        rev_graph[going_state_num][init_going_state] = [];
      }
      rev_graph[going_state_num][init_going_state] =
        rev_graph[going_state_num][init_going_state].concat(chars);
    }
  }

  if (accept_nodes.size === 0) {
    throw new Error("accept node must exist");
  }
  const accept_nodes_array = [...accept_nodes];
  if (accept_nodes_array.length !== 1) {
    throw new Error("the size of accept nodes must be one");
  }

  let eq_i: number = 0;
  let lt_i: number = 0;
  let and_i: number = 0;
  let multi_or_i: number = 0;

  const range_checks: number[][][] = new Array(256);
  for (let i = 0; i < 256; i++) {
    range_checks[i] = new Array(256);
  }
  const eq_checks: number[] = new Array(256);
  const multi_or_checks1: Record<string, number> = {};
  const multi_or_checks2: Record<string, number> = {};

  let lines: string[] = [];
  lines.push(`\tfor (var i = 0; i < num_bytes; i++) {`);

  // const uppercase = new Set(Array.from("ABCDEFGHIJKLMNOPQRSTUVWXYZ").map(c => c.charCodeAt()));
  // const lowercase = new Set(Array.from("abcdefghijklmnopqrstuvwxyz").map(c => c.charCodeAt()));
  // const digits = new Set(Array.from("0123456789").map(c => c.charCodeAt()));
  // const symbols1 = new Set([":", ";", "<", "=", ">", "?", "@"].map(c => c.charCodeAt()));
  // const symbols2 = new Set(["[", "\\", "]", "^", "_", "`"].map(c => c.charCodeAt()));
  // const symbols3 = new Set(["{", "|", "}", "~"].map(c => c.charCodeAt()));
  lines.push(`\t\tstate_changed[i] = MultiOR(${N - 1});`);

  for (let i = 1; i < N; i++) {
    const outputs: number[] = [];
    // let is_negates = [];
    for (const prev_i of Object.keys(rev_graph[i])) {
      const prev_i_num = Number(prev_i);
      const k = rev_graph[i][prev_i_num];
      k.sort((a, b) => Number(a) - Number(b));
      const eq_outputs: [string, number][] = [];
      let vals: Set<number> = new Set(k);
      // let is_negate = false;
      // if (vals.has(0xff)) {
      //     vals.delete(0xff);
      //     is_negate = true;
      // }
      if (vals.size === 0) {
        continue;
      }
      // if (is_negate === true) {
      //     for (let another_i = 1; another_i < N; another_i++) {
      //         if (i === another_i) {
      //             continue;
      //         }
      //         if (rev_graph[another_i][prev_i] === null) {
      //             continue;
      //         }
      //         const another_vals = new Set(rev_graph[another_i][prev_i]);
      //         if (another_vals.size === 0) {
      //             continue;
      //         }
      //         for (let another_val of another_vals) {
      //             vals.add(another_val);
      //         }
      //     }
      // }
      const min_maxes: [number, number][] = [];
      let cur_min: number = k[0];
      let cur_max: number = k[0];

      for (let idx = 1; idx < k.length; ++idx) {
        if (cur_max === k[idx]) {
          continue;
        } else if (cur_max + 1 === k[idx]) {
          cur_max += 1;
        } else {
          if (cur_max - cur_min >= 16) {
            min_maxes.push([cur_min, cur_max]);
          }
          cur_min = k[idx];
          cur_max = k[idx];
        }
      }

      if (cur_max - cur_min >= 16) {
        min_maxes.push([cur_min, cur_max]);
      }
      for (const min_max of min_maxes) {
        for (let code = min_max[0]; code <= min_max[1]; ++code) {
          vals.delete(code);
        }
      }

      // for (let subsets of [
      //     [digits, 47, 58],
      //     [symbols1, 57, 65],
      //     [uppercase, 64, 91],
      //     [symbols2, 90, 97],
      //     [lowercase, 96, 123],
      //     [symbols3, 122, 127]
      // ]) {
      //     const subset = subsets[0];
      //     const min = subsets[1];
      //     const max = subsets[2];
      //     if (vals.isSuperset(subset)) {
      //         vals.difference(subset);
      //         if (min_maxs.length == 0) {
      //             min_maxs.push([min, max]);
      //         } else {
      //             const last = min_maxs[min_maxs.length - 1];
      //             if (last[1] - 1 == min) {
      //                 min_maxs[min_maxs.length - 1][1] = max;
      //             } else {
      //                 min_maxs.push([min, max]);
      //             }
      //         }
      //     }
      // }

      for (let min_max of min_maxes) {
        const min: number = min_max[0];
        const max: number = min_max[1];
        if (range_checks[min][max] === undefined) {
          lines.push(`\t\tlt[${lt_i}][i] = LessEqThan(8);`);
          lines.push(`\t\tlt[${lt_i}][i].in[0] <== ${min};`);
          lines.push(`\t\tlt[${lt_i}][i].in[1] <== in[i];`);

          lines.push(`\t\tlt[${lt_i + 1}][i] = LessEqThan(8);`);
          lines.push(`\t\tlt[${lt_i + 1}][i].in[0] <== in[i];`);
          lines.push(`\t\tlt[${lt_i + 1}][i].in[1] <== ${max};`);

          lines.push(`\t\tand[${and_i}][i] = AND();`);
          lines.push(`\t\tand[${and_i}][i].a <== lt[${lt_i}][i].out;`);
          lines.push(`\t\tand[${and_i}][i].b <== lt[${lt_i + 1}][i].out;`);

          eq_outputs.push(["and", and_i]);
          range_checks[min][max] = [lt_i, and_i];
          lt_i += 2;
          and_i += 1;
        } else {
          let [_, and_i] = range_checks[min][max];
          eq_outputs.push(["and", and_i]);
        }
      }
      for (let code of vals) {
        if (eq_checks[code] === undefined) {
          lines.push(`\t\teq[${eq_i}][i] = IsEqual();`);
          lines.push(`\t\teq[${eq_i}][i].in[0] <== in[i];`);
          lines.push(`\t\teq[${eq_i}][i].in[1] <== ${code};`);
          eq_outputs.push(["eq", eq_i]);
          eq_checks[code] = eq_i;
          eq_i += 1;
        } else {
          eq_outputs.push(["eq", eq_checks[code]]);
        }
      }

      lines.push(`\t\tand[${and_i}][i] = AND();`);
      lines.push(`\t\tand[${and_i}][i].a <== states[i][${prev_i}];`);
      if (eq_outputs.length === 1) {
        // if (is_negate) {
        //     lines.push(`\t\tand[${and_i}][i].b <== 1 - ${eq_outputs[0][0]}[${eq_outputs[0][1]}][i].out;`);
        // } else {
        //     lines.push(`\t\tand[${and_i}][i].b <== ${eq_outputs[0][0]}[${eq_outputs[0][1]}][i].out;`);
        // }
        lines.push(
          `\t\tand[${and_i}][i].b <== ${eq_outputs[0][0]}[${eq_outputs[0][1]}][i].out;`
        );
      } else if (eq_outputs.length > 1) {
        const eq_outputs_key: string = JSON.stringify(eq_outputs);
        if (multi_or_checks1[eq_outputs_key] === undefined) {
          lines.push(
            `\t\tmulti_or[${multi_or_i}][i] = MultiOR(${eq_outputs.length});`
          );
          for (let output_i = 0; output_i < eq_outputs.length; output_i++) {
            lines.push(
              `\t\tmulti_or[${multi_or_i}][i].in[${output_i}] <== ${eq_outputs[output_i][0]}[${eq_outputs[output_i][1]}][i].out;`
            );
          }
          // if (is_negate) {
          //     lines.push(`\t\tand[${and_i}][i].b <== 1 - multi_or[${multi_or_i}][i].out;`);
          // } else {
          //     lines.push(`\t\tand[${and_i}][i].b <== multi_or[${multi_or_i}][i].out;`);
          // }
          lines.push(
            `\t\tand[${and_i}][i].b <== multi_or[${multi_or_i}][i].out;`
          );
          multi_or_checks1[eq_outputs_key] = multi_or_i;
          multi_or_i += 1;
        } else {
          lines.push(
            `\t\tand[${and_i}][i].b <== multi_or[${multi_or_checks1[eq_outputs_key]}][i].out;`
          );
        }
      }

      outputs.push(and_i);
      and_i += 1;
    }

    if (outputs.length === 1) {
      lines.push(`\t\tstates[i+1][${i}] <== and[${outputs[0]}][i].out;`);
    } else if (outputs.length > 1) {
      const outputs_key: string = JSON.stringify(outputs);
      if (multi_or_checks2[outputs_key] === undefined) {
        lines.push(
          `\t\tmulti_or[${multi_or_i}][i] = MultiOR(${outputs.length});`
        );
        for (let output_i = 0; output_i < outputs.length; output_i++) {
          lines.push(
            `\t\tmulti_or[${multi_or_i}][i].in[${output_i}] <== and[${outputs[output_i]}][i].out;`
          );
        }
        lines.push(`\t\tstates[i+1][${i}] <== multi_or[${multi_or_i}][i].out;`);
        multi_or_checks2[outputs_key] = multi_or_i;
        multi_or_i += 1;
      } else {
        lines.push(
          `\t\tstates[i+1][${i}] <== multi_or[${multi_or_checks2[outputs_key]}][i].out;`
        );
      }
    }

    lines.push(`\t\tstate_changed[i].in[${i - 1}] <== states[i+1][${i}];`);
  }

  lines.push(`\t\tstates[i+1][0] <== 1 - state_changed[i].out;`);
  lines.push("\t}");

  const declarations: string[] = [];
  declarations.push(`pragma circom 2.1.5;\n`);
  declarations.push(
    `include "@zk-email/zk-regex-circom/circuits/regex_helpers.circom";\n`
  );
  // declarations.push(`pragma circom 2.1.5;\ninclude "@zk-email/circuits/regexes/regex_helpers.circom";\n`);
  declarations.push(`// regex: ${regex_str.replace(/\n/g, "\\n")}`);
  declarations.push(`template ${template_name}(msg_bytes) {`);
  declarations.push(`\tsignal input msg[msg_bytes];`);
  declarations.push(`\tsignal output out;\n`);
  declarations.push(`\tvar num_bytes = msg_bytes+1;`);
  declarations.push(`\tsignal in[num_bytes];`);
  declarations.push(`\tin[0]<==255;`);
  declarations.push(`\tfor (var i = 0; i < msg_bytes; i++) {`);
  declarations.push(`\t\tin[i+1] <== msg[i];`);
  declarations.push(`\t}\n`);
  if (eq_i > 0) {
    declarations.push(`\tcomponent eq[${eq_i}][num_bytes];`);
  }
  if (lt_i > 0) {
    declarations.push(`\tcomponent lt[${lt_i}][num_bytes];`);
  }
  if (and_i > 0) {
    declarations.push(`\tcomponent and[${and_i}][num_bytes];`);
  }
  if (multi_or_i > 0) {
    declarations.push(`\tcomponent multi_or[${multi_or_i}][num_bytes];`);
  }
  declarations.push(`\tsignal states[num_bytes+1][${N}];`);
  declarations.push(`\tcomponent state_changed[num_bytes];`);
  declarations.push("");

  const init_code: string[] = [];
  init_code.push(`\tstates[0][0] <== 1;`);
  init_code.push(`\tfor (var i = 1; i < ${N}; i++) {`);
  init_code.push(`\t\tstates[0][i] <== 0;`);
  init_code.push("\t}");
  init_code.push("");

  lines = declarations.concat(init_code).concat(lines);

  const accept_node: number = accept_nodes_array[0];
  const accept_lines = [""];
  accept_lines.push("\tcomponent final_state_result = MultiOR(num_bytes+1);");
  accept_lines.push("\tfor (var i = 0; i <= num_bytes; i++) {");
  accept_lines.push(
    `\t\tfinal_state_result.in[i] <== states[i][${accept_node}];`
  );
  accept_lines.push("\t}");
  accept_lines.push("\tout <== final_state_result.out;");

  lines = lines.concat(accept_lines);
  let string: string = lines.reduce((res, line) => res + line + "\n", "");
  return string;
}

// Commented these two out as they're only used by the code that's also commented out

// Set.prototype.isSuperset = function (subset) {
//     if (this.size === 0) {
//         return false;
//     }
//     for (var elem of subset) {
//         if (!this.has(elem)) {
//             return false;
//         }
//     }
//     return true;
// }

// Set.prototype.difference = function (setB) {
//     for (let elem of setB) {
//         this.delete(elem)
//     }
// }

File: packages/compiler/src/gen_circom.rs
use serde::{Deserialize, Serialize};
use std::collections::{BTreeSet, HashMap};

fn gen_circom_allstr(graph_json: Vec<GraphNode>, template_name: &str, regex_str: &str) -> String {
    let n = graph_json.len();
    let mut rev_graph = HashMap::<u32, HashMap<u32, Vec<u8>>>::new();
    let mut to_init_graph = vec![];
    let mut init_going_state: Option<u32> = None;

    for i in 0..n {
        rev_graph.insert(i as u32, HashMap::new());
        to_init_graph.push(vec![]);
    }

    let mut accept_nodes = BTreeSet::<u32>::new();

    for i in 0..n {
        let node = &graph_json[i];
        for (k, v) in &node.edges {
            let value: Vec<u8> = v.iter().cloned().collect();
            rev_graph
                .get_mut(k)
                .unwrap()
                .insert(i as u32, value.clone());

            if i == 0 {
                // if let Some(index) = value.iter().position(|&x| x == 94) {
                //     init_going_state = Some(*v);
                //     rev_graph.get_mut(&v).unwrap().get_mut(&i).unwrap()[index] = 255;
                // }

                for j in rev_graph.get(&k).unwrap().get(&(i as u32)).unwrap() {
                    if *j == 255 {
                        continue;
                    }
                    to_init_graph[*k as usize].push(*j);
                }
            }
        }

        if node.r#type == "accept" {
            accept_nodes.insert(i as u32);
        }
    }

    if let Some(init_going_state) = init_going_state {
        for (going_state, chars) in to_init_graph.iter().enumerate() {
            if chars.is_empty() {
                continue;
            }

            if rev_graph
                .get_mut(&(going_state as u32))
                .unwrap()
                .get_mut(&init_going_state)
                .is_none()
            {
                rev_graph
                    .get_mut(&(going_state as u32))
                    .unwrap()
                    .insert(init_going_state, vec![]);
            }

            rev_graph
                .get_mut(&(going_state as u32))
                .unwrap()
                .get_mut(&init_going_state)
                .unwrap()
                .extend_from_slice(chars);
        }
    }

    if accept_nodes.is_empty() {
        panic!("Accept node must exist");
    }

    let accept_nodes_array: Vec<u32> = accept_nodes.into_iter().collect();

    if accept_nodes_array.len() != 1 {
        panic!("The size of accept nodes must be one");
    }

    let mut eq_i = 0;
    let mut lt_i = 0;
    let mut and_i = 0;
    let mut multi_or_i = 0;

    let mut range_checks = vec![vec![None; 256]; 256];
    let mut eq_checks = vec![None; 256];
    let mut multi_or_checks1 = HashMap::<String, usize>::new();
    let mut multi_or_checks2 = HashMap::<String, usize>::new();

    let mut lines = vec![];

    lines.push("\tfor (var i = 0; i < num_bytes; i++) {".to_string());
    lines.push(format!("\t\tstate_changed[i] = MultiOR({});", n - 1));

    for i in 1..n {
        let mut outputs = vec![];

        for (prev_i, k) in rev_graph.get(&(i as u32)).unwrap().iter() {
            let prev_i_num = *prev_i;
            let mut k = k.clone();
            k.sort();

            let mut eq_outputs = vec![];
            let mut vals = k.clone().into_iter().collect::<BTreeSet<u8>>();

            if vals.is_empty() {
                continue;
            }

            let mut min_maxes = vec![];
            let mut cur_min = k[0];
            let mut cur_max = k[0];

            for idx in 1..k.len() {
                if cur_max == k[idx] {
                    continue;
                } else if cur_max + 1 == k[idx] {
                    cur_max += 1;
                } else {
                    if cur_max - cur_min >= 16 {
                        min_maxes.push((cur_min, cur_max));
                    }
                    cur_min = k[idx];
                    cur_max = k[idx];
                }
            }

            if cur_max - cur_min >= 16 {
                min_maxes.push((cur_min, cur_max));
            }

            for min_max in &min_maxes {
                for code in min_max.0..=min_max.1 {
                    vals.remove(&code);
                }
            }

            for min_max in &min_maxes {
                let min = min_max.0;
                let max = min_max.1;

                if range_checks[min as usize][max as usize].is_none() {
                    lines.push(format!("\t\tlt[{}][i] = LessEqThan(8);", lt_i));
                    lines.push(format!("\t\tlt[{}][i].in[0] <== {};", lt_i, min));
                    lines.push(format!("\t\tlt[{}][i].in[1] <== in[i];", lt_i));
                    lines.push(format!("\t\tlt[{}][i] = LessEqThan(8);", lt_i + 1));
                    lines.push(format!("\t\tlt[{}][i].in[0] <== in[i];", lt_i + 1));
                    lines.push(format!("\t\tlt[{}][i].in[1] <== {};", lt_i + 1, max));
                    lines.push(format!("\t\tand[{}][i] = AND();", and_i));
                    lines.push(format!("\t\tand[{}][i].a <== lt[{}][i].out;", and_i, lt_i));
                    lines.push(format!(
                        "\t\tand[{}][i].b <== lt[{}][i].out;",
                        and_i,
                        lt_i + 1
                    ));
                    eq_outputs.push(("and", and_i));
                    range_checks[min as usize][max as usize] = Some((lt_i, and_i));
                    lt_i += 2;
                    and_i += 1;
                } else {
                    if let Some((_, and_i)) = range_checks[min as usize][max as usize] {
                        eq_outputs.push(("and", and_i));
                    }
                }
            }

            for code in &vals {
                if eq_checks[*code as usize].is_none() {
                    lines.push(format!("\t\teq[{}][i] = IsEqual();", eq_i));
                    lines.push(format!("\t\teq[{}][i].in[0] <== in[i];", eq_i));
                    lines.push(format!("\t\teq[{}][i].in[1] <== {};", eq_i, code));
                    eq_outputs.push(("eq", eq_i));
                    eq_checks[*code as usize] = Some(eq_i);
                    eq_i += 1;
                } else {
                    if let Some(eq_i) = eq_checks[*code as usize] {
                        eq_outputs.push(("eq", eq_i));
                    }
                }
            }

            lines.push(format!("\t\tand[{}][i] = AND();", and_i));
            lines.push(format!(
                "\t\tand[{}][i].a <== states[i][{}];",
                and_i, prev_i_num
            ));

            if eq_outputs.len() == 1 {
                lines.push(format!(
                    "\t\tand[{}][i].b <== {}[{}][i].out;",
                    and_i, eq_outputs[0].0, eq_outputs[0].1
                ));
            } else if eq_outputs.len() > 1 {
                let eq_outputs_key = serde_json::to_string(&eq_outputs).unwrap();

                if multi_or_checks1.get(&eq_outputs_key).is_none() {
                    lines.push(format!(
                        "\t\tmulti_or[{}][i] = MultiOR({});",
                        multi_or_i,
                        eq_outputs.len()
                    ));

                    for (output_i, (eq_type, eq_i)) in eq_outputs.iter().enumerate() {
                        lines.push(format!(
                            "\t\tmulti_or[{}][i].in[{}] <== {}[{}][i].out;",
                            multi_or_i, output_i, eq_type, eq_i
                        ));
                    }

                    lines.push(format!(
                        "\t\tand[{}][i].b <== multi_or[{}][i].out;",
                        and_i, multi_or_i
                    ));
                    multi_or_checks1.insert(eq_outputs_key, multi_or_i);
                    multi_or_i += 1;
                } else {
                    if let Some(multi_or_i) = multi_or_checks1.get(&eq_outputs_key) {
                        lines.push(format!(
                            "\t\tand[{}][i].b <== multi_or[{}][i].out;",
                            and_i, multi_or_i
                        ));
                    }
                }
            }

            outputs.push(and_i);
            and_i += 1;
        }

        if outputs.len() == 1 {
            lines.push(format!(
                "\t\tstates[i+1][{}] <== and[{}][i].out;",
                i, outputs[0]
            ));
        } else if outputs.len() > 1 {
            let outputs_key = serde_json::to_string(&outputs).unwrap();

            if multi_or_checks2.get(&outputs_key).is_none() {
                lines.push(format!(
                    "\t\tmulti_or[{}][i] = MultiOR({});",
                    multi_or_i,
                    outputs.len()
                ));

                for (output_i, and_i) in outputs.iter().enumerate() {
                    lines.push(format!(
                        "\t\tmulti_or[{}][i].in[{}] <== and[{}][i].out;",
                        multi_or_i, output_i, and_i
                    ));
                }

                lines.push(format!(
                    "\t\tstates[i+1][{}] <== multi_or[{}][i].out;",
                    i, multi_or_i
                ));
                multi_or_checks2.insert(outputs_key, multi_or_i);
                multi_or_i += 1;
            } else {
                if let Some(multi_or_i_) = multi_or_checks2.get(&outputs_key) {
                    lines.push(format!(
                        "\t\tstates[i+1][{}] <== multi_or[{}][i].out;",
                        i, multi_or_i_
                    ));
                }
            }
        }

        lines.push(format!(
            "\t\tstate_changed[i].in[{}] <== states[i+1][{}];",
            i - 1,
            i
        ));
    }

    // println!("multi_or_i: {}", multi_or_i);

    lines.push("\t\tstates[i+1][0] <== 1 - state_changed[i].out;".to_string());

    let mut declarations = vec![];
    declarations.push("pragma circom 2.1.5;\n".to_string());
    declarations
        .push("include \"@zk-email/zk-regex-circom/circuits/regex_helpers.circom\";\n".to_string());
    declarations.push(format!(
        "// regex: {}",
        regex_str.replace("\n", "\\n").replace("\r", "\\r"),
    ));
    declarations.push(format!("template {}(msg_bytes) {{", template_name));
    declarations.push("\tsignal input msg[msg_bytes];".to_string());
    declarations.push("\tsignal output out;\n".to_string());
    declarations.push("\tvar num_bytes = msg_bytes+1;".to_string());
    declarations.push("\tsignal in[num_bytes];".to_string());
    declarations.push("\tin[0]<==255;".to_string());
    declarations.push("\tfor (var i = 0; i < msg_bytes; i++) {".to_string());
    declarations.push("\t\tin[i+1] <== msg[i];".to_string());
    declarations.push("\t}\n".to_string());

    if eq_i > 0 {
        declarations.push(format!("\tcomponent eq[{}][num_bytes];", eq_i));
    }

    if lt_i > 0 {
        declarations.push(format!("\tcomponent lt[{}][num_bytes];", lt_i));
    }

    if and_i > 0 {
        declarations.push(format!("\tcomponent and[{}][num_bytes];", and_i));
    }

    if multi_or_i > 0 {
        println!("multi_or_i: {}", multi_or_i);
        declarations.push(format!("\tcomponent multi_or[{}][num_bytes];", multi_or_i));
    }

    declarations.push(format!("\tsignal states[num_bytes+1][{}];", n));
    declarations.push("\tcomponent state_changed[num_bytes];\n".to_string());

    let mut init_code = vec![];
    init_code.push("\tstates[0][0] <== 1;".to_string());
    init_code.push(format!("\tfor (var i = 1; i < {}; i++) {{", n));
    init_code.push("\t\tstates[0][i] <== 0;".to_string());
    init_code.push("\t}\n".to_string());

    let mut final_code = declarations
        .into_iter()
        .chain(init_code)
        .chain(lines)
        .collect::<Vec<String>>();
    final_code.push("\t}".to_string());

    let accept_node = accept_nodes_array[0];
    let mut accept_lines = vec![];

    accept_lines.push("".to_string());
    accept_lines.push("\tcomponent final_state_result = MultiOR(num_bytes+1);".to_string());
    accept_lines.push("\tfor (var i = 0; i <= num_bytes; i++) {".to_string());
    accept_lines.push(format!(
        "\t\tfinal_state_result.in[i] <== states[i][{}];",
        accept_node
    ));
    accept_lines.push("\t}".to_string());
    accept_lines.push("\tout <== final_state_result.out;".to_string());

    final_code.extend(accept_lines);

    final_code.join("\n")
}

fn main() {
    let json_str = r#"[
        {
          "type": "",
          "state": 0,
          "edges": {
            "1": [
              97,
              98
            ]
          }
        },
        {
          "type": "accept",
          "state": 1,
          "edges": {
            "1": [
              97,
              98
            ]
          }
        }
      ]"#;

    let graph_json: Vec<GraphNode> = serde_json::from_str(json_str).unwrap();

    println!("{:?}", graph_json);

    let circom_str = gen_circom_allstr(graph_json, "test", "((\r\n)|^)subject:[^\r\n]+\r\n");

    println!("{}", circom_str);
}
